### PR TITLE
Putting everything together

### PR DESCRIPTION
**Pull request by gkousouris**
Putting all our building blocks together:
image

Added README for setup explanation. TODO: set up requirements.txt with all dependencies.

### CODE REVIEW

File: logic.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+    repertoire = '1. e4 (1. d4 e5) e5 2. Nf3 Nc6'
+    game = '1. e4 c5 2. Na3 Nc6 3. b4 b5'
+    assert find_first_deviation(repertoire, game, False)['message'] == 'Blunder, you should have played e5 according to your prep.', find_first_deviation(repertoire, game, False)
+
+    repertoire = '1. e4 (1. d4 e5 2. dxe5) e5 2. Nf3 Nc6'
+    game = '1. d4 e5 2. dxe5 Nf6 a4 b5'
+    assert find_first_deviation(repertoire, game, False)['message'] == 'Prep ran out, you played Nf6, novelty.', find_first_deviation(repertoire, game, False)
+
+# testcases()
```
Comment (gkousouris) [inline logic.py:72]: Created testcases to show behaviour of this function

File: logic.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import io
+import chess.pgn
+from utils import get_sharpest_lines_from_fen
+
+def get_san(node, move):
+    return node.board().san(move)
+
+# TODO: Be able to infer is_user_white in the future from repertoire / get request.
```
Comment (gkousouris) [inline logic.py:8]: Can't think of a straightforward way of figuring out if user is white/black apart from directly asking them.

File: utils.py [hunk 3]
```diff
@@ -13,17 +14,18 @@
 # pip install chess
 import chess.pgn
 
+stockfish = Stockfish('/usr/local/bin/stockfish')
 
 # IMPORT GAMES
 def get_moves_from_gameID(gameID):
-
```
Comment (gkousouris) [inline utils.py:19]: my code editor doesn't like tabs xD

File: utils.py [hunk 4]
```diff
@@ -177,59 +184,66 @@ def get_mistake_blunder_likelihood_from_fen(fen, mistake_threshold=200, blunder_
...
         print("Not enough games in Opening Explorer")
         print("")
-    
+
     return
 
 
-def get_sharpest_lines_from_fen(fen="rn1qk1nr/pp3pbp/4p1p1/2ppP3/3P4/2N2B1P/PPP2PP1/R1BQK2R w KQkq - 0 9", mistake_threshold=150, blunder_threshold=400, verbose=False):
-    # Put thresholds negative when analysing white moves 
+def get_sharpest_lines_from_fen(fen="rn1qk1nr/pp3pbp/4p1p1/2ppP3/3P4/2N2B1P/PPP2PP1/R1BQK2R w KQkq - 0 9", mistake_threshold=150, blunder_threshold=400, verbose=False, dev_limit=100):
```
Comment (gkousouris) [inline utils.py:230]: added a dev_limit=2 to speed up development.

File: logic.py (hunk 2 above)
Comment (jewe6889) [reply logic.py:8]: Perhaps we can do it when the user is importing the PGN instead

File: utils.py (hunk 4 above)
Comment (jewe6889) [reply utils.py:230]: Great idea

**Review by jewe6889 — APPROVED** [review]

**Comment by jewe6889** [comment]
Merging it to test the Flask server :)

**Comment by jewe6889** [comment]
> In the first terminal, run the following command to start the front-end:
> http-server

@gkousouris I'm not sure how to make that work. Do I require anything else to start the front-end?

**Comment by jewe6889** [comment]
Just thinking, we could actually name this project "Where did I go wrong in the opening?", as it's basically what it is haha.

**Comment by gkousouris** [comment]
> > In the first terminal, run the following command to start the front-end:
> > http-server
> 
> @gkousouris I'm not sure how to make that work. Do I require anything else to start the front-end?

@jewe6889 You need to run `npm install -g http-server` for that to work. (also may need to install npm 😅 )

**Comment by jewe6889** [comment]
Thx I'll try. Btw, been thinking of having something like that...

image

**Comment by jewe6889** [comment]
I managed to make the server work, I'm just stuck with importing Stockfish (tried to change the path to mine, but still didn't work). Should we simply add it to this repo?

**Comment by gkousouris** [comment]
> I managed to make the server work, I'm just stuck with importing Stockfish (tried to change the path to mine, but still didn't work). Should we simply add it to this repo?

What's the issue? Are you using python3?

**Comment by jewe6889** [comment]
> > I managed to make the server work, I'm just stuck with importing Stockfish (tried to change the path to mine, but still didn't work). Should we simply add it to this repo?
> 
> What's the issue? Are you using python3?

Yeah I just don't have any usr/local/bin folder on Windows, so I was trying to put my own folder, but it still didn't recognise it for some reason.

**Comment by gkousouris** [comment]
> > > I managed to make the server work, I'm just stuck with importing Stockfish (tried to change the path to mine, but still didn't work). Should we simply add it to this repo?
> 
> > 
> 
> > What's the issue? Are you using python3?
> 
> 
> 
> Yeah I just don't have any usr/local/bin folder on Windows, so I was trying to put my own folder, but it still didn't recognise it for some reason.

What does "which stockfish" say? You have to point it to the binary. I did pip3 install stockfish and then point it to that

**Pull request merged**